tmt/checks/watchdog.py

```diff
@@ -0,0 +1,470 @@
+
+            logger.fail('connection refused')
+
+            guest_context.ssh_ping_failures += 1
+
+            report_progress(log,
+                            'ssh-ping',
+                            [
+                                '# connection refused',
+                                f'# failed {guest_context.ssh_ping_failures}'
+                                f' of {self.ssh_ping_threshold} allowed',
+                                ],
+                            command_output=ncat_output)
+
+        def _success(ncat_output: str) -> None:
+            """ Handle successfull response """
+
+            logger.verbose('Received successful response to SSH ping.', level=2)
+
+            report = [
+                '# successfull response'
+                ]
+
+            if guest_context.ssh_ping_failures != 0:
+                report.append(f'# replenished failure budget back to {self.ssh_ping_threshold}')
+
+            guest_context.ssh_ping_failures = 0
+
+            report_progress(
+                log,
+                'ssh-ping',
+                report,
+                command_output=ncat_output
+                )
+
+        try:
+            assert invocation.guest.primary_address is not None  # narrow type
+
+            output = tmt.utils.Command('nc',
+                                       '-zv',
+                                       invocation.guest.primary_address,
+                                       str(invocation.guest.port or 22)) .run(cwd=Path.cwd(),
+                                                                              logger=logger)
+
+            _success(output.stderr or '')
+
+        except tmt.utils.RunError as exc:
+            if exc.returncode == 1:
+                _fail_connection_refused(exc.stderr or '')
+
+            else:
+                _fail_unknown('\n'.join(render_run_exception_streams(exc.stdout, exc.stderr)))
+
+        logger.debug(
+            f'failed {guest_context.ssh_ping_failures}'
+            f' of {self.ssh_ping_threshold} allowed')
+
+        if guest_context.ssh_ping_failures >= self.ssh_ping_threshold:
+            logger.fail(f'exhausted {self.ssh_ping_threshold} SSH ping attempts')
+
+            self.notify(invocation, logger)
+
+
+@provides_check('watchdog')
+class Watchdog(CheckPlugin[WatchdogCheck]):
+    """
+    Take various actions when guest becomes unresponsive.
+
+    Watchdog runs selected probes every now and then, and when a given
+    number of `probes` fail, watchdog would run one or more of the
+    predefined `actions`.
+
+    Check comes with two probes, "ping" and "SSH ping", and single
+    action, "reboot".
+
+    * "ping" uses the classic ICMP echo to check whether the guest is
+      still up and running,
+    * "SSH ping" tries to establish SSH connection,
+    * "reboot" action issues a hard reboot of the guest.
+
+    Each probe has a "budged" of allowed failures, and when it runs out,
```

Comment:
```suggestion
    Each probe has a "budget" of allowed failures, and when it runs out,
```

Reply:
Fixed in 410efea7b260e7f0fced4e99b3609a0b67881e03